File: backend/nestjs/apps/core-service/src/instructions/instructions.controller.ts
import { Controller } from '@nestjs/common';
import { GrpcMethod } from '@nestjs/microservices';
import { InstructionsService } from './instructions.service';
import { UpdateInstructionDto } from './dto/update-instruction.dto';

@Controller()
export class InstructionsController {
  constructor(private readonly instructionsService: InstructionsService) {}

  @GrpcMethod('InstructionsService', 'FindInstructionByUserId')
  findByUserId(data: { user_id: string }) {
    return this.instructionsService.findByUserId(data.user_id);
  }

  @GrpcMethod('InstructionsService', 'CreateOrUpdateInstruction')
  createOrUpdate(data: { instruction: string; user_id: string }) {
    const updateInstructionDto: UpdateInstructionDto = {
      instruction: data.instruction,
      userId: data.user_id,
    };
    return this.instructionsService.createOrUpdate(updateInstructionDto);
  }
}

File: backend/java/src/main/java/ai/snippetquiz/core_service/controller/InstructionsController.java
package ai.snippetquiz.core_service.controller;

import ai.snippetquiz.core_service.dto.request.CreateOrUpdateInstructionRequest;
import ai.snippetquiz.core_service.dto.response.InstructionResponse;
import ai.snippetquiz.core_service.service.InstructionsService;
import ai.snippetquiz.core_service.util.Constants;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.UUID;

@RestController
@RequestMapping("/instructions")
@RequiredArgsConstructor
public class InstructionsController {
    
    private final InstructionsService instructionsService;
    
    @GetMapping
    public ResponseEntity<InstructionResponse> findByUserId(
            @RequestHeader(Constants.USER_ID_HEADER) String userId) {
        InstructionResponse response = instructionsService.findByUserId(UUID.fromString(userId));
        return ResponseEntity.ok(response);
    }
    
    @PutMapping
    public ResponseEntity<InstructionResponse> createOrUpdate(
            @RequestHeader(Constants.USER_ID_HEADER) String userId,
            @Valid @RequestBody CreateOrUpdateInstructionRequest request) {
        InstructionResponse response = instructionsService.createOrUpdate(UUID.fromString(userId), request);
        return ResponseEntity.ok(response);
    }
}